Add reboot service for obihai

absurdist81 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Registers a service, obihai.reboot, which can be called in an automation.  This new reboot service depends on approval of [https://github.com/dshokouhi/pyobihai/pull/5](https://github.com/dshokouhi/pyobihai/pull/5).


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

absurdist81 committed:
Define and register reboot service

This new service depends on approval of https://github.com/dshokouhi/pyobihai/pull/5

absurdist81 committed:
reboot service description

absurdist81 committed:
fix newlines

absurdist81 committed:
remove whitespace from blank lines

absurdist81 committed:
fix newlines

absurdist81 committed:
new line after function

absurdist81 committed:
fix quotation marks

frenck commented:
Converted this PR to draft for now. Please mark it ready for review when that point is reached. 👍

absurdist81 committed:
Black formatting

dshokouhi commented:
New package published to pypi, going to remove the upstream label now

absurdist81 committed:
Bump required pyobihai version

version 1.3.0 required for reboot service

dshokouhi commented on homeassistant/components/obihai/manifest.json:
```diff
@@ -2,6 +2,6 @@
   "name": "Obihai",
   "documentation": "https://www.home-assistant.io/integrations/obihai",
-  "requirements": ["pyobihai==1.2.3"],
+  "requirements": ["pyobihai==1.3.0"],
```
In addition to changing the version here you also need to perform the second bullet point: https://developers.home-assistant.io/docs/development_checklist

absurdist81 replied:
Thanks!

absurdist81 replied:
I tried updating the pyobihai version in requirements_all.txt manually, and there was still a failed check.  I ran the gen_requirements_all script and uploaded the resulting requirements_all.txt, but that had many conflicts, so I changed it back.  I don't have a proper development environment for HA and github.  I've just made been coding on my "production" server (my raspberry pi hassio setup) and uploading the files to github.com.

absurdist81 replied:
The remaining failed check (pylint exits with code 4) seems to be unrelated to requirements_all.txt.  Does it just not like the comment?

absurdist81 committed:
bump pyobihai to 1.3.0

pyobihai 1.3.0 required for reboot service

absurdist81 committed:
bump pyobihai to 1.3.0

absurdist81 committed:
bump pyobihai to 1.3.0

absurdist81 committed:
remove comment

dshokouhi commented on homeassistant/components/obihai/sensor.py:
```diff
@@ -48,8 +48,6 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
         _LOGGER.error("Invalid credentials")
         return
 
-    """Define and register reboot service"""
```
The reason why it complained is because the docstring is supposed to be after the line `def _reboot(call):` also docstrings need to end with a period `.`

absurdist81 replied:
I appreciate the explanation.  I don't usually program in Python, so I'm learning.

dshokouhi commented on homeassistant/components/obihai/sensor.py:
```diff
@@ -48,6 +48,13 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
         return
 
+    def _reboot(call):
+        success = pyobihai.call_reboot()
```
Let's add back the doc string in the correct place, also make sure to end it with a period like the rest of the functions on this page.  I am not 100% sure if docstrings are required by the core team but since we have it on every function we should make sure it matches.

absurdist81 replied:
Sure thing.

dshokouhi reviewed (changes requested):
Lets put back the docstring in the correct location as the rest of the functions on this page have one too.  It may also be a core requirement but I am not certain.

absurdist81 committed:
docstring for reboot function

dshokouhi reviewed (approved):
Looks good to me

ludeeus commented on homeassistant/components/obihai/sensor.py:
```diff
@@ -48,6 +48,14 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+        if not success:
+            _LOGGER.error("Failed to reboot")
+
+    hass.services.register("obihai", "reboot", _reboot)
```
When services are defined in platforms it should use the entity service helper <https://developers.home-assistant.io/docs/dev_101_services#entity-services>

absurdist81 replied:
I tried to do that in my earliest attempts at defining the service, but I did not get it working.  I don't have what I coded or the error messages, but I believe that entity_platform.current_platform.get() was None.  Perhaps because no entities/devices are created in in the obihai integration.  I could try again if I find the time.

absurdist81 replied:
I have working code with the entity service helper now.

ludeeus commented on homeassistant/components/obihai/services.yaml:
```diff
@@ -0,0 +1,3 @@
+reboot:
+  description: Sends command to reboot obihai device
```
```suggestion
  description: Sends reboot command to the Obihai device
```

absurdist81 committed:
Update homeassistant/components/obihai/services.yaml

Co-authored-by: Joakim Sørensen <hi@ludeeus.dev>

absurdist81 committed:
register reboot as entity service

Use entity service helper instead of hass.services.register. Also changed setup_platform to async_setup_platform and add_entities to async_add_entities.

absurdist81 committed:
reboot service requires entity_id

absurdist81 committed:
cleanup for black

dshokouhi commented:
@absurdist81 just a FYI some of the CI checks have not passed, you may want to check the errors and if necessary force-push to get the checks to pass.  Usually helps speed up reviews :)

ludeeus commented on homeassistant/components/obihai/sensor.py:
```diff
@@ -32,7 +34,7 @@
 
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
```
You are now doing sync I/O in async context, this needs to be handled.
Have a look here <https://developers.home-assistant.io/docs/asyncio_working_with_async#calling-sync-functions-from-async>
